Spigot-Server-Patches/0698-Add-ExitPortalCreateEvent.patch

```diff
@@ -0,0 +1,49 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: ravkr <rav_kr@interia.pl>
+Date: Fri, 26 Mar 2021 02:25:05 +0100
+Subject: [PATCH] Add ExitPortalCreateEvent
+
+
+diff --git a/src/main/java/net/minecraft/world/level/dimension/end/EnderDragonBattle.java b/src/main/java/net/minecraft/world/level/dimension/end/EnderDragonBattle.java
+index 0ab6319aa3e4e1f5679f37be36999ca56ca2484c..0887972626ffc67a4df49755e804996c066e8db1 100644
+--- a/src/main/java/net/minecraft/world/level/dimension/end/EnderDragonBattle.java
++++ b/src/main/java/net/minecraft/world/level/dimension/end/EnderDragonBattle.java
+@@ -58,6 +58,8 @@ import net.minecraft.world.phys.AxisAlignedBB;
+ import org.apache.logging.log4j.LogManager;
+ import org.apache.logging.log4j.Logger;
+ import io.papermc.paper.event.block.DragonEggFormEvent; // Paper - DragonEggFormEvent
++import io.papermc.paper.event.world.ExitPortalCreateEvent; // Paper - ExitPortalCreateEvent
++import org.bukkit.Location; // Paper - ExitPortalCreateEvent
+ 
+ public class EnderDragonBattle {
+ 
+@@ -442,14 +444,27 @@ public class EnderDragonBattle {
+     }
+ 
+     public void generateExitPortal(boolean flag) {
+-        WorldGenEndTrophy worldgenendtrophy = new WorldGenEndTrophy(flag);
+-
++        // Paper start - Call ExitPortalCreateEvent and use results
+         if (this.exitPortalLocation == null) {
+             for (this.exitPortalLocation = this.world.getHighestBlockYAt(HeightMap.Type.MOTION_BLOCKING_NO_LEAVES, WorldGenEndTrophy.a).down(); this.world.getType(this.exitPortalLocation).a(Blocks.BEDROCK) && this.exitPortalLocation.getY() > this.world.getSeaLevel(); this.exitPortalLocation = this.exitPortalLocation.down()) {
+                 ;
+             }
+         }
+ 
++        if (exitPortalLocation.getY() < 1) exitPortalLocation.setY(1); // fix portal generating at y=-1 when the world is empty (skyblock generators etc.); maybe also MC-93185
++
++        Location location = new Location(this.world.getWorld(), exitPortalLocation.getX(), exitPortalLocation.getY(), exitPortalLocation.getZ());
++        ExitPortalCreateEvent exitPortalEvent = new ExitPortalCreateEvent(location, flag);
++        org.bukkit.Bukkit.getPluginManager().callEvent(exitPortalEvent);
++        if (exitPortalEvent.isCancelled()) return;
```

Comment:
This line, and the one above it can be combined into a single line. `if (!exitPortalEvent.callEvent()) { return; }